### PR TITLE
Address concerns of weak-into-raw

### PR DESCRIPTION
**Pull request by vorner**
This should address the standing concerns in https://github.com/rust-lang/rust/issues/60728#issuecomment-612525616.

I've still left the ability to create a new dangling pointer from `null`, as I feel like this is the natural behaviour to expect, but I'm fine removing that too. I've modified the documentation to allow the `as_ptr` or `into_ptr` to return whatever garbage in case of a dangling pointer. I've also removed the guarantee to be able to do `from_raw(as_ptr)` from the documentation (but it would still work right now).

I've renamed the method and added implementations for `Rc`/`Arc`.

I've also tried if I can just „enable“ unsized types. I believe the current interface is compatible with them. But the inner implementation will be a bit challenging ‒ I can't use the `data_offset` as is used by `Rc` or `Arc` because it AFAIK „touches“ (creates a reference to) the live value of `T` ‒ and in case of `Weak`, it might be completely bogus or already dead ‒ so that would be UB.

`./x.py test tidy` is completely mad on my own system all over the code base :-(. I'll just hope it goes through CI, or will fix as necessary.

Is it OK if I ask @Amanieu for review, as the concerns are from you?

~r @Amanieu

### CODE REVIEW

**Comment by Centril** [comment]
cc @RalfJung

File: src/liballoc/rc.rs [hunk 1]
```diff
@@ -569,9 +569,33 @@ impl<T: ?Sized> Rc<T> {
...
     #[stable(feature = "rc_raw", since = "1.17.0")]
     pub fn into_raw(this: Self) -> *const T {
+        let ptr = Self::as_ptr(&this);
+        mem::forget(this);
+        ptr
+    }
+
+    /// Provides a raw pointer to the data.
+    ///
+    /// Without changing any counts or consuming the value. The pointer is valid for as long there
```
Comment (RalfJung) [inline src/liballoc/rc.rs:579]: The first "sentence" here is not a proper sentence.

File: src/liballoc/sync.rs [hunk 2]
```diff
@@ -566,9 +566,33 @@ impl<T: ?Sized> Arc<T> {
...
     #[stable(feature = "rc_raw", since = "1.17.0")]
     pub fn into_raw(this: Self) -> *const T {
+        let ptr = Self::as_ptr(&this);
+        mem::forget(this);
+        ptr
+    }
+
+    /// Provides a raw pointer to the data.
+    ///
+    /// Without changing any counts or consuming the value. The pointer is valid for as long as
```
Comment (Amanieu) [inline src/liballoc/sync.rs:576]: Also here.

**Comment by Amanieu** [comment]
r=me once the issues with the doc comments are fixed.

File: src/liballoc/rc.rs (hunk 1 above)
Comment (vorner) [reply src/liballoc/rc.rs]: I meant it as continuation of the caption, but you're right, it's not really smooth transition. Changed.

**Comment by vorner** [comment]
Thanks. I've forced-pushed it directly, but the only change is the rewording of the comment.

**Pull request merged**